File: 백준/Silver/9012. 괄호/괄호.py
import sys
from collections import deque
input = sys.stdin.readline

n = int(input())
stack = deque()
flag = False
for _ in range(n):
    input_list = input().strip()
    for ps in input_list:
        if ps == '(':
            stack.append(ps)
        elif ps == ')':
            if not stack:
                # 비어 있으면 바로 return
                flag = True
                break
            if stack[-1] == '(':
                stack.pop()
            else:
                stack.append(ps)

    if len(stack) > 0 or flag:
        print("NO")
        stack.clear()
        flag = False
    else:
        print("YES")

File: 백준/Silver/9012. 괄호/괄호.kt

fun main() {
    val br = System.`in`.bufferedReader()
    val sb = StringBuilder()
    val n = br.readLine().toInt()

    var stack = mutableListOf<Char>()
    for(i in 0 until n) {
        val ps = br.readLine()

        for(c in ps){
            if (c == '('){
                stack.add(c)
            }else if(c==')'){
                if(stack.isEmpty()){
                    stack.add(c)
                    break
                }
                stack.removeLast()
            }
        }

        if(stack.isNotEmpty()){
            sb.append("NO\n")
            stack.clear()
        }else {
            sb.append("YES\n")
        }
    }
    print(sb)
}
